### PR TITLE
Feature/chapter 3

### PR DESCRIPTION
**Pull request by cbeyer42**

### CODE REVIEW

File: 3-Network-Configuration/3.1-helper.bash [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        t=1
+      fi
+    done
+    [ "$t" -eq 0 ] && echo "Wireless is not enabled"
+  else
+    echo "Wireless is not enabled"
+  fi
+}
+
+function test_3_1_1_ipv6_grub() {
```
Comment (robertauer) [inline 3-Network-Configuration/3.1-helper.bash]: The name of this function does not describe what it is doing or what kind of response the user can expect from it.
I would suggest to either rename the function or integrate it into the test that uses it.

File: 3-Network-Configuration/3.1-helper.bash [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+function test_3_1_2() {
```
Comment (robertauer) [inline 3-Network-Configuration/3.1-helper.bash]: Maybe put this function into its own file as it is done in chapter 6, e.g. `3.1.2.sh`

File: 3-Network-Configuration/3.2-Network-Parameters-Host-Only.bats [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    [ "$output" = "net.ipv4.ip_forward = 0" ]
+    run bash -c "grep -E -s \"^\s*net\.ipv4\.ip_forward\s*=\s*1\" /etc/sysctl.conf /etc/sysctl.d/*.conf /usr/lib/sysctl.d/*.conf /run/sysctl.d/*.conf"
+    [ "$status" -ne 0 ]
+    [ "$output" = "" ]
+    run bash -c "sysctl net.ipv6.conf.all.forwarding"
+    [ "$status" -eq 0 ]
+    [ "$output" = "net.ipv6.conf.all.forwarding = 0" ]
+    run bash -c "grep -E -s \"^\s*net\.ipv6\.conf\.all\.forwarding\s*=\s*1\" /etc/sysctl.conf /etc/sysctl.d/*.conf /usr/lib/sysctl.d/*.conf /run/sysctl.d/*.conf"
+    [ "$status" -ne 0 ]
+    [ "$output" = "" ]
```
Comment (robertauer) [inline 3-Network-Configuration/3.2-Network-Parameters-Host-Only.bats]: These tests can only be executed when IPv6 is active. Maybe you can check this beforehand via the `check_ip_v6` function of the `IPv6-helper.bash` like you did in 3.3.1

File: 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    [ "$status" -eq 0 ]
+    # Check if the desired output line is active in any of the conf files
+    local CONF_ALL_FILE_CORRECT=0
+    while IFS= read -r line; do
+        if [[ "$line" == *":net.ipv4.conf.all.accept_source_route = 0" ]]; then
+            CONF_ALL_FILE_CORRECT=1
+        fi
+    done <<< "$output"
+    [ $CONF_ALL_FILE_CORRECT -eq 1 ]
+    run bash -c "grep "net\.ipv4\.conf\.default\.accept_source_route" /etc/sysctl.conf /etc/sysctl.d/*"
```
Comment (robertauer) [inline 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats]: Missing escape of quotes
```suggestion
    run bash -c "grep \"net\.ipv4\.conf\.default\.accept_source_route\" /etc/sysctl.conf /etc/sysctl.d/*"
```

File: 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        if [[ "$line" == *":net.ipv4.conf.default.accept_source_route = 0" ]]; then
+            CONF_DEFAULT_FILE_CORRECT=1
+        fi
+    done <<< "$output"
+    [ $CONF_DEFAULT_FILE_CORRECT -eq 1 ]
+
+    run check_ip_v6
+    [ $status -eq 0 ]
+    if [[ "$output" == *"*** IPv6 is enabled on the system ***"* ]]; then
+        echo "INFO: IPv6 is enabled on the system"
```
Comment (robertauer) [inline 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats]: Is this debug output and can be removed?

File: 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+            CONF_FILE_CORRECT=1
+        fi
+    done <<< "$output"
+    [ $CONF_FILE_CORRECT -eq 1 ]
+
+    # checks for ipv6
+    run check_ip_v6
+    [ $status -eq 0 ]
+    if [[ "$output" == *"*** IPv6 is enabled on the system ***"* ]]; then
+        echo "INFO: IPv6 is enabled on the system"
```
Comment (robertauer) [inline 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats]: Is this debug output and can be removed?

File: 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+@test "3.3.9 Ensure IPv6 router advertisements are not accepted (Automated)" {
+    run check_ip_v6
+    [ $status -eq 0 ]
+    if [[ "$output" != *"*** IPv6 is enabled on the system ***"* ]]; then
+        skip "*** IPv6 is not enabled on the system ***"
+    fi
+    
+    echo "INFO: IPv6 is enabled on the system"
```
Comment (robertauer) [inline 3-Network-Configuration/3.3-Network-Parameters-Host-And-Router.bats]: Is this debug output and can be removed?

File: 3-Network-Configuration/3.4-Uncommon-Network-Protocols.bats [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    run bash -c "modprobe -n -v dccp"
+    [ "$status" -eq 0 ]
+    [ "$output" = "install /bin/true " ]
+    run bash -c "lsmod | grep dccp"
+    [ "$status" -ne 0 ]
+    [ "$output" = "" ]
+}
+
+@test "3.4.2 Ensure SCTP is disabled (Automated)" {
+    run bash -c "modprobe -n -v sctp"
```
Comment (robertauer) [inline 3-Network-Configuration/3.4-Uncommon-Network-Protocols.bats]: Please change this to the command listed in the CIS document (`modprobe -n -v sctp | grep -E '(sctp|install)'`)

File: 3-Network-Configuration/3.4-Uncommon-Network-Protocols.bats [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    run bash -c "modprobe -n -v rds"
+    [ "$status" -eq 0 ]
+    [ "$output" = "install /bin/true " ]
+    run bash -c "lsmod | grep rds"
+    [ "$status" -ne 0 ]
+    [ "$output" = "" ]
+}
+
+@test "3.4.4 Ensure TIPC is disabled (Automated)" {
+    (modprobe -n -v tipc | grep "install /bin/true")
```
Comment (robertauer) [inline 3-Network-Configuration/3.4-Uncommon-Network-Protocols.bats]: Please change this to the format all other tests use and use the command listed in the CIS document, e.g.
```suggestion
    run bash -c "modprobe -n -v tipc | grep -E '(tipc|install)'"
    [ "$status" -eq 0 ]
    [ "$output" = "install /bin/true " ]
```

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+load IPv6-helper
+
+# 3.5.1 Configure UncomplicatedFirewall
+
+@test "3.5.1.1 Ensure ufw is installed (Automated)" {
+    [[ $(dpkg -s ufw | grep 'Status: install') == "Status: install ok installed" ]]
+}
+
+@test "3.5.1.2 Ensure iptables-persistent is not installed with ufw (Automated)" {
+    run dpkg-query -s iptables-persistent
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats]:
```suggestion
    run bash -c "dpkg-query -s iptables-persistent"
```

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+
+@test "3.5.1.3 Ensure ufw service is enabled (Automated)" {
+    [[ $(systemctl is-enabled ufw) == "enabled" ]]
+    [[ $(ufw status | grep Status) == "Status: active" ]]
+}
+
+@test "3.5.1.4 Ensure ufw loopback traffic is configured (Automated)" {
+    run bash -c "ufw status verbose"
+    [ "$status" -eq 0 ]
+    [[ "$output" == *"Anywhere on lo"*"ALLOW IN"*"Anywhere"* ]]
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats]: This statement doesn't make sure that all the strings are in the same line. Maybe use the `$lines` array: https://bats-core.readthedocs.io/en/latest/writing-tests.html

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+@test "3.5.1.7 Ensure ufw default deny firewall policy (Automated)" {
+    run bash -c "ufw status verbose | grep -i default"
+    [ "$status" -eq 0 ]
+    [[ "$output" == *"deny (incoming)"* || "$output" == *"reject (incoming)"* ]]
+    [[ "$output" == *"deny (outgoing)"* || "$output" == *"reject (outgoing)"* ]]
+    [[ "$output" == *"deny (routed)"* \
+     || "$output" == *"reject (routed)"* \
+     || "$output" == *"disabled (routed)"* ]]
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats]: Where does `disabled` come from? It is not mentioned in the CIS document.

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+# 3.5.2 Configure nftables
+
+@test "3.5.2.1 Ensure nftables is installed (Automated)" {
+    run bash -c "dpkg -s nftables | grep 'Status: install'"
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats]: This does not match the command from the CIS document (`dpkg-query -s nftables | grep 'Status: install ok installed'`)

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    [[ "$output" == "Status: install ok installed" ]]
+}
+
+@test "3.5.2.2 Ensure ufw is uninstalled or disabled with nftables (Automated)" {
+    run bash -c "dpkg -s ufw | grep 'dpkg-query'"
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats]: This does not match the command from the CIS document (`dpkg-query -s ufw | grep 'Status: install ok installed'`)

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+@test "3.5.2.4 Ensure a nftables table exists (Automated)" {
+    run bash -c "nft list tables"
+    [ "$status" -eq 0 ]
+}
+
+@test "3.5.2.5 Ensure nftables base chains exist (Automated)" {
+    run bash -c "nft list ruleset | grep 'hook input'"
+    [ "$status" -eq 0 ]
+    [[ "$output" == *"type filter hook input priority filter;"* ]]
```
Comment (robertauer) [inline 3-Network-Configuration/3.5-Firewall-Configuration.bats:95]: This does not match the line from the CIS document (`type filter hook input priority 0;`)
Also check the other two in this test

**Review by robertauer — CHANGES_REQUESTED** [review]

File: 3-Network-Configuration/3.5-Firewall-Configuration.bats (hunk 15 above)
Comment (cbeyer42) [reply 3-Network-Configuration/3.5-Firewall-Configuration.bats:95]: This must be an error in the CIS document. `priority 0` will be outputted as `priority filter`.
Here you can see a list of possible priority settings: https://wiki.nftables.org/wiki-nftables/index.php?title=Configuring_chains&oldid=208#Base_chain_priority

Comment (robertauer) [reply 3-Network-Configuration/3.5-Firewall-Configuration.bats:95]: Yes, it seems so. ok.

**Review by robertauer — APPROVED** [review]

**Pull request merged**